File: shared/bin/docker-container-update.py
#!/usr/bin/env python


import json
import os
from collections import defaultdict
from pathlib import Path
from typing import NamedTuple

from hl_helpers import Color, Log
from shellrunner import ShellCommandError, X

os.environ["SHELLRUNNER_SHOW_OUTPUT"] = "False"
os.environ["SHELLRUNNER_SHOW_COMMAND"] = "False"


class ContainerDetails(NamedTuple):
    image_name: str
    service_name: str
    compose_file_path: Path
    build: bool
    is_up_to_date: bool


def get_running_containers() -> list[tuple[str, ...]]:
    containers = X(
        "docker ps --format '{{.ID}}::{{.Image}}'",
    ).out.splitlines()
    return [tuple(container.split("::")) for container in containers]


def get_container_details(container_id: str, image: str) -> ContainerDetails | None:
    try:
        image_name = X(f"docker inspect {container_id}" + """ --format='{{.Config.Image}}'""").out
        service_name = X(
            f"docker inspect {container_id}" """ --format '{{ index .Config.Labels "com.docker.compose.service" }}'""",
        ).out
        compose_file = X(
            f"docker inspect {container_id}"
            """ --format '{{ index .Config.Labels "com.docker.compose.project.config_files" }}'""",
        ).out
        compose_file_path = Path(compose_file).resolve(strict=True)

        compose_config = X(f"docker compose -f {compose_file_path} config --format json").out
        compose_config = json.loads(compose_config)
        build = "build" in compose_config["services"][service_name]

        is_up_to_date = False
        if not build:
            image_id = X(f"docker inspect {container_id}" + " --format '{{.Image}}'").out
            image_digest = X(f"docker image inspect {image_id}" + " --format '{{.RepoDigests}}'").out
            remote_digest = X(f"regctl image digest {image}").out
            is_up_to_date = remote_digest in image_digest

        return ContainerDetails(image_name, service_name, compose_file_path, build, is_up_to_date)
    except ShellCommandError as e:
        Log.warn(f"An error occured while getting the details for container: {image} ({container_id})")
        Log.error(e.out)


def build_compose_files_to_containers(
    running_containers: list[tuple[str, ...]],
) -> defaultdict[Path, list[ContainerDetails]]:
    compose_files_to_containers: defaultdict[Path, list[ContainerDetails]] = defaultdict(list)
    for container_id, image in running_containers:
        container_details = get_container_details(container_id, image)
        if container_details is None:
            continue

        compose_files_to_containers[container_details.compose_file_path].append(container_details)

    return compose_files_to_containers


def print_container_update_details(container: ContainerDetails) -> None:
    status = ""
    if container.build:
        status = Color.warn("Build")
    elif container.is_up_to_date:
        status = "Up to date"
    else:
        status = Color.success("Update found")

    print(f"{Color.info(container.service_name)}[{container.image_name}]: {status}")


def update_container(container: ContainerDetails) -> None:
    compose_file_dir = container.compose_file_path.parent
    if container.build:
        Log.info(f"Building {container.service_name}[{container.image_name}]...")
        X(
            [
                f"cd {compose_file_dir}",
                f"docker compose -f {container.compose_file_path} build --no-cache --pull {container.service_name}",
                f"docker compose -f {container.compose_file_path} up -d {container.service_name}",
            ],
        )
    elif not container.is_up_to_date:
        Log.info(f"Updating {container.service_name}[{container.image_name}]...")
        X(
            [
                f"cd {compose_file_dir}",
                f"docker pull {container.image_name}",
                f"docker compose -f {container.compose_file_path} up -d {container.service_name}",
            ],
        )


def main() -> None:
    Log.info("Getting details for running containers...")
    running_containers = get_running_containers()
    compose_files_to_containers = build_compose_files_to_containers(running_containers)

    for compose_file_path, containers in compose_files_to_containers.items():
        Log.notice(f"\n== {compose_file_path} ==")
        for container in containers:
            print_container_update_details(container)
        for container in containers:
            update_container(container)


if __name__ == "__main__":
    main()

File: shared/bin/docker-container-update.ts
#!/usr/bin/env bun

import path from "node:path";
import { $, type Shell, type ShellExpression } from "bun";

interface BunShellOptions {
	cwd?: string;
	env?: Record<string, string | undefined>;
	quiet?: boolean;
}

/**
 * This is a wrapper around Bun shell to make it easier to create new instances.
 */
class BunShell {
	#shell: Shell;
	#cwd = "";
	#env: Record<string, string | undefined> = {};
	#quiet = false;

	constructor(options?: BunShellOptions) {
		const { cwd, env, quiet } = options ?? {};

		this.#shell = new $.Shell();
		this.#shell.throws(true);

		this.cwd = cwd ?? process.cwd();
		this.env = env ?? Bun.env;
		this.quiet = quiet ?? false;
	}

	$(strings: TemplateStringsArray, ...expressions: ShellExpression[]) {
		if (this.quiet) return this.#shell(strings, ...expressions).quiet();
		return this.#shell(strings, ...expressions);
	}

	get cwd() {
		return this.#cwd;
	}

	set cwd(path: string) {
		this.#shell.cwd(path);
		this.#cwd = path;
	}

	get env() {
		return this.#env;
	}

	set env(env: Record<string, string | undefined>) {
		this.#shell.env(env);
		this.#env = env;
	}

	get quiet() {
		return this.#quiet;
	}

	set quiet(quiet: boolean) {
		this.#quiet = quiet;
	}
}

interface ContainerDetails {
	imageName: string;
	serviceName: string;
	composeFilePath: string;
	build: boolean;
	isUpToDate: boolean;
}

const shell = new BunShell({ quiet: true });

async function getRunningContainers() {
	const containers = (
		await shell.$`docker ps --format '{{.ID}}::{{.Image}}'`.text()
	)
		.trim()
		.split("\n");
	return containers.map((container) => container.split("::"));
}

async function getContainerDetails(
	containerId: string,
	image: string,
): Promise<ContainerDetails | undefined> {
	try {
		const imageName = (
			await shell.$`docker inspect ${containerId} --format='{{.Config.Image}}'`.text()
		).trim();
		const serviceName = (
			await shell.$`docker inspect ${containerId} --format '{{ index .Config.Labels "com.docker.compose.service" }}'`.text()
		).trim();
		const composeFile = (
			await shell.$`docker inspect ${containerId} --format '{{ index .Config.Labels "com.docker.compose.project.config_files" }}'`.text()
		).trim();
		const composeFilePath = path.resolve(composeFile);
		const composeFileExists = await Bun.file(composeFilePath).exists();
		if (!composeFileExists) {
			throw new Error(`Compose file not found: ${composeFilePath}`);
		}

		const composeConfigJson = (
			await shell.$`docker compose -f ${composeFilePath} config --format json`.text()
		).trim();
		const composeConfig = JSON.parse(composeConfigJson);

		const build = Object.hasOwn(composeConfig.services[serviceName], "build");

		let isUpToDate = false;
		if (!build) {
			const imageId = (
				await shell.$`docker inspect ${containerId} --format '{{.Image}}'`.text()
			).trim();
			const imageDigest = (
				await shell.$`docker image inspect ${imageId} --format '{{.RepoDigests}}'`.text()
			).trim();
			const remoteDigest = (
				await shell.$`regctl image digest ${image}`.text()
			).trim();
			isUpToDate = imageDigest.includes(remoteDigest);
		}

		return { imageName, serviceName, composeFilePath, build, isUpToDate };
	} catch (error) {
		console.error(
			`An error occured while getting the details for container: ${image} (${containerId})`,
		);
		console.error(error);
	}
}

async function buildComposeFilesToContainers(
	runningContainers: string[][],
): Promise<Map<string, ContainerDetails[]>> {
	const composeFilesToContainers: Map<string, ContainerDetails[]> = new Map();
	for (const [containerId, image] of runningContainers) {
		const containerDetails = await getContainerDetails(containerId, image);
		if (!containerDetails) continue;
		const composeFilePath = containerDetails.composeFilePath;
		composeFilesToContainers.set(composeFilePath, [
			...(composeFilesToContainers.get(composeFilePath) || []),
			containerDetails,
		]);
	}
	return composeFilesToContainers;
}

function printContainerUpdateDetails(container: ContainerDetails) {
	let status = "";
	if (container.build) {
		status = "Build";
	} else if (container.isUpToDate) {
		status = "Up to date";
	} else {
		status = "Update found";
	}
	console.log(`${container.serviceName}[${container.imageName}]: ${status}`);
}

async function updateContainer(container: ContainerDetails) {
	const composeFileDir = path.dirname(container.composeFilePath);
	if (container.build) {
		console.log(`Building ${container.serviceName}[${container.imageName}]...`);
		await shell.$`cd ${composeFileDir} && docker compose -f ${container.composeFilePath} build --no-cache --pull ${container.serviceName} && docker compose -f ${container.composeFilePath} up -d ${container.serviceName}`;
	} else if (!container.isUpToDate) {
		console.log(`Updating ${container.serviceName}[${container.imageName}]...`);
		await shell.$`cd ${composeFileDir} && docker pull ${container.imageName} && docker compose -f ${container.composeFilePath} up -d ${container.serviceName}`;
	}
}

async function main() {
	console.log("Getting details for running containers...");
	const runningContainers = await getRunningContainers();
	const composeFilesToContainers =
		await buildComposeFilesToContainers(runningContainers);

	for (const [composeFilePath, containers] of composeFilesToContainers) {
		console.log(`\n== ${composeFilePath} ==`);
		for (const container of containers) {
			printContainerUpdateDetails(container);
		}
		for (const container of containers) {
			await updateContainer(container);
		}
	}
}

await main();
